FIPS/fip-0032.md

```diff
@@ -21,290 +21,426 @@ replaces: N/A
-
-- `block_open` (extern-traversing): `block_open_base + block_open_per_byte * len`
-- `block_read`: `block_read_base + block_memcpy_per_byte * len`
-- `block_write`: `block_create_base + block_memcpy_per_byte * len`
-- `block_link` (extern-traversing): `block_link_base + block_link_per_byte * storage_gas_multiplire * len`
-- `block_stat`: `block_stat` (fixed cost)
+The cost of syscall-driven context switches varies depending on the syscall, its
+parameters, the amount of work it performs (thus invalidating CPU caches and
+instruction pipeline), and other factors. In the future, we may tabulate syscall
+gas depending on the complexity of the syscall, its arguments, its impact on CPU
+optimizations, and more.
 
-### Extern-traversing syscall fee revision
+### Extern gas
 
-Some syscalls cannot be resolved entirely within FVM space. Such syscalls
-need to traverse the "extern" (FVM<>node) boundary to access client
-functionality.
+Some syscalls cannot be resolved entirely within FVM space. Such syscalls need
+to traverse the _extern_ (FVM <> Node) boundary to access client functionality.
 
 Depending on the languages of the client and the FVM implementation, traversing
-this boundary may involve foreign-function interface (FFI) mechanics.
-
-That is indeed the case with the reference client (Lotus) and the reference FVM
-(ref-fvm). Not only is this the worst case scenario, but it's the most common
-scenario today given the client distribution of the network.
+this boundary may involve foreign-function interface (FFI) mechanics. That is
+indeed the case with the reference client (Lotus) and the reference FVM
+(ref-fvm). This is the most common scenario today given the client distribution
+of the network.
 
-In accordance to the design principle of high-execution fidelity, we propose
-that the gas fees account for the cost of the FFI boundary.
+The associated overheads are:
 
-The associated costs are:
-
-- data structure translation between representations, including serialization
-  costs.
+- data structure translation between representations, including potential
+  serialization overheads.
 - context switch.
 - the memory overhead on both sides of the interface.
-- the amortized (opportunity) cost of dedicating OS threads.
+- the amortized opportunity cost of dedicating OS threads to handle FFI calls.
 - the amortized garbage collection or deallocation costs once the extern
   concludes.
 - function lookup and dispatch.
 
 **Pricing formulae**
 
-The pricing formulae for extern-traversing syscalls are:
+Extern gas is priced at 1.5x syscall gas. It is equivalent to a flat 5250
+execution units, and therefore it costs 21000 gas. Extern gas is additive to
+syscall gas.
+
+```
+p_extern_gas := 21000
+```
+
+The following syscalls are extern-traversing syscalls, and therefore are subject
+to extern gas:
+
+- `rand::get_chain_randomness`.
+- `rand::get_beacon_randomness`.
+- `crypto::verify_consensus_fault`.
+- `ipld::block_open`.
+
```

Comment:
future considerations:

- We can drop the randomness gas by constructing the FVM with 2*finality worth of VRF proofs & beacons. This is 225 bytes of data that we can pass _very_ efficiently. See https://github.com/filecoin-project/ref-fvm/issues/525.
- We may, eventually, move consensus fault verification entirely on-chain.

(but I'm not sure if we need to mention these)